pylint/checkers/refactoring/refactoring_checker.py

```diff
@@ -1393,7 +1393,43 @@ def _check_unnecessary_comprehension(self, node):
         else:
             return
         if expr_list == target_list != []:
-            self.add_message("unnecessary-comprehension", node=node)
+            if isinstance(node.parent, astroid.DictComp) and isinstance(
+                utils.safe_infer(node.iter), astroid.objects.DictItems
+            ):
+                self.add_message(
+                    "unnecessary-comprehension",
+                    node=node,
+                    args=(f"{node.iter.func.expr.as_string()}",),
+                )
+                return
+            if (
+                isinstance(node.parent, astroid.ListComp)
+                and isinstance(utils.safe_infer(node.iter), astroid.List)
+            ) or (
+                isinstance(node.parent, astroid.SetComp)
+                and isinstance(utils.safe_infer(node.iter), astroid.Set)
+            ):
+                self.add_message(
+                    "unnecessary-comprehension",
+                    node=node,
+                    args=(f"{node.iter.as_string()}",),
+                )
+                return
```

Comment:
I _could_ group these 2 if blocks into a single, but that throws pylint error of having too much boolean conditions in a single if block -- therefore I had to split them up

Reply:
Maybe define `args` in the conditional and then use it in add_message later ? Something like this:

```suggestion
            if isinstance(node.parent, astroid.DictComp) and isinstance(
                utils.safe_infer(node.iter), astroid.objects.DictItems
            ):
                args=(f"{node.iter.func.expr.as_string()}",)
            if (
                isinstance(node.parent, astroid.ListComp)
                and isinstance(utils.safe_infer(node.iter), astroid.List)
            ) or (
                isinstance(node.parent, astroid.SetComp)
                and isinstance(utils.safe_infer(node.iter), astroid.Set)
            ):
                    args=(f"{node.iter.as_string()}",)
           if args:
               self.add_message("unnecessary-comprehension", node=node, args=args)
```

Reply:
You can disable pylint warning if they make the code worst too :)

Reply:
the suggested code is much cleaner, I think I'll go with that!